### DISCUSSION TITLE
OSX Sandbox - Network issues

### DISCUSSION THREAD

**Discussion by IcemarkUK** [axmolengine/axmol#1669] [Q&A] [answered]
My OSX build is setup with code signing and entitlements, as discussed [here](https://github.com/axmolengine/axmol/discussions/1662).

Sandbox is now working.

The entitlement file looks like it's being picked up.

<img width="989" alt="Screenshot 2024-02-11 at 19 18 41" src="https://github.com/axmolengine/axmol/assets/8301416/44cec96a-d7bd-4f02-bcb3-2c95b986e453">

And the entitlement file has `com.apple.security.network.client` set to true.

https://github.com/ChilliHugger/The-Lords-Of-Midnight/blob/393d8c6c315e67096d78209794e3b673dd384307/main/midnight/proj.ios_mac/mac/midnight.entitlements#L1-L10

In my cocos build I can access a URL and download from it.

```
    auto request = new HttpRequest();
    request->setRequestType(HttpRequest::Type::GET);
    request->setUrl("https://thelordsofmidnight.com/version/lom_v2.0_osx.cfg");
    request->setResponseCallback([&,callback](HttpClient* client, HttpResponse* response)
    {
        try
        {
            if(response->isSucceed())
            {
// ... do stuff
            }
        } catch (const std::exception& e) {
        }
    });

    HttpClient::getInstance()->send(request);
    request->release();
```
 
In my axmol build the response fails and I see this in the logs...
```
  [yasio][1707678408336][index: 0] start query thelordsofmidnight.com...
  [yasio][1707678408336][index: 0] connecting server thelordsofmidnight.com(188.64.190.46):443...
  [yasio][1707678408337][index: 0] connect server thelordsofmidnight.com(188.64.190.46):443 failed, ec=1, detail:Operation not permitted
```

If I remove the entitlement file and disable code signing and set to run locally - it works fine.

**Comment by halx99** [comment]
The error reported by os, check does follow link helpful?
https://github.com/flutter/flutter/issues/47606

**Reply by IcemarkUK** [reply]
I'll go hunting for the OS error...

But the link above doesn't help as I already have the network client entitlement set, and it doesn't suggest anything else.

**Comment by rh101** [comment]
Is there any chance this is not entirely network related, but that something is attempting to read or write from a location that can't be accessed without specific entitlements being enabled?

**Reply by IcemarkUK** [reply]
Hard to say - this is the link - https://thelordsofmidnight.com/version/lom_v2.0_osx.cfg
It's hittable outside of the app - and it's hittable in cocos version with sandbox and entitlements on - and it's hittable in axmol buildIf I turn the sandbox signing off... But it's not hittable in axmol version with Sandbox and entitlements on. I'm not aware of anything else that might need to be turn on... Unless of course the new library requires incoming connections to also be enabled...

**Reply by rh101** [reply]
So it is most likely an issue in the current HttpClient implementation.

Any chance you can try using the `ax::network::Downloader` to retrieve the file to see if that works? If it's not the preferred method, or if it just doesn't work, you could try using another HTTP library for the time being, until we discover what the issue is with the Axmol HttpClient.  If that's something you would want to do, then one that I've personally had success with is [libhv](https://github.com/ithewei/libhv), which is very simple to integrate and use.

**Reply by halx99** [reply]
A more simple method to verify network issue is use raw bsd socket try to create tcp connection.  i.e use `yasio/xxsocket.hpp`

**Reply by halx99** [reply]
```cpp
yasio::xxsocket sock;
int ret = sock.xpconnect_n("xxx.com", 443, std::chrono::seconds(3));
bool succed = ret == 0;
```

**Reply by IcemarkUK** [reply]
Ok, I'll give that a try later this evening.

**Reply by IcemarkUK** [reply]
So the yasio::xxsocket appeared to work. I also grabbed the CurlTest code, dropped it in to my app, and pointed at my URL - it worked fine.

I managed to get it to work - but not ideally - it turns out that the network code does indeed need the SERVER entitlement (`com_apple_security_network_server`) as well as the CLIENT entitlement (`com.apple.security.network.client`). That seems a little excessive to me.

**Reply by rh101** [reply]
@halx99 Is the issue with HttpClient? `com.apple.security.network.server` is only required for listening to incoming connections, so it's a little odd that it has to be enabled in this case, where the client is simply opening outgoing connections.

**Reply by halx99** [reply]
does there is a minimal demo I can reproduce this issue?

**Reply by halx99** [reply]
this branch https://github.com/ChilliHugger/The-Lords-Of-Midnight/tree/AXMOL-port ?

**Reply by halx99** [reply]
I get this error on branch `AXMOL-port`
![image](https://github.com/axmolengine/axmol/assets/6977319/d3ca7e1a-f8dd-4c89-91c5-d1b9f0f1478c)

**Reply by rh101** [reply]
This needs to be passed to it: `-xc '-DTME=LOM'` 

On Windows, this command works for me:  `axmol build -p win32 -c -xc '-DTME=LOM'`

**Reply by halx99** [reply]
![image](https://github.com/axmolengine/axmol/assets/6977319/d43925e0-fba4-4372-9819-bef0d0889cbb)

**Reply by rh101** [reply]
Only on Windows, you may have an issue with these lines:

`ax_sync_target_res(${APP_NAME} LINK_TO ${APP_RES_DIR} FOLDERS ${content_folder} SYM_LINK 1)`
and
`set_property(TARGET ${APP_NAME} PROPERTY VS_DEBUGGER_WORKING_DIRECTORY "${content_folder}")`

because there are 2 resource folders defined:
```
set(content_folder
    "${CMAKE_CURRENT_SOURCE_DIR}/Resources"
    "${CMAKE_CURRENT_SOURCE_DIR}/Resources-${APP_NAME}"
    )
```

The only way I got it working on Windows is to remove the Content symlink in the output binary folder, create a Content folder, then copy the contents of `Resources` and `Resources-midnight` to it,

**Reply by rh101** [reply]
Sorry didn't notice you were building for MacOS; never-mind the info regarding Windows.

**Reply by halx99** [reply]
![image](https://github.com/axmolengine/axmol/assets/6977319/f1bd96db-e2c4-4772-b5f0-046fc70b9df7)
does it require enrolled apple developer account?

**Reply by rh101** [reply]
It may, but I'm not certain.  I couldn't figure out how to create a sandbox version of the application without signing it, so couldn't do anything to test it in that mode.

@IcemarkUK Is there a way to create sandbox version of the application without the developer account?

**Reply by halx99** [reply]
I disable code sign
![image](https://github.com/axmolengine/axmol/assets/6977319/4de5a3ad-15d5-4109-8146-051b17539ed6)


then get this compile error:
![image](https://github.com/axmolengine/axmol/assets/6977319/f4e200a1-fbbe-4183-87f6-0e4f56c07e4c)

**Reply by halx99** [reply]
build success, but when I run it:

![image](https://github.com/axmolengine/axmol/assets/6977319/fc6bca96-9bcc-4e22-a3b0-7d6a8004af3a)

NOTE: I also enable `-DTESTS=YES`

**Reply by rh101** [reply]
Without `-DTESTS`, I can build and run it successfully, but I have another issue. It's using full screen mode when it starts, and it's blanking out my monitor.  Not sure what is going on there, but I had to VNC into the Mac, then switch it in the main menu of the app to "Small" resolution to recover the monitor output.

I can see this in the log now too:
`[yasio][1707793735515][index: 0] connect server thelordsofmidnight.com(188.64.190.46):443 failed, ec=1, detail:Operation not permitted`

**Reply by halx99** [reply]
![image](https://github.com/axmolengine/axmol/assets/6977319/2b17e056-5bf6-4977-9e86-36b63f7cb983)
Just change screen mode to window. 

I think I know what happy, will fix later

**Reply by rh101** [reply]
Thanks for that tip!

In my case, it's a very specific setup.  144Hz monitor running through KVM which only supports 120Hz, so the monitor connected to the Mac is set to 120Hz.  When the app loads in full screen, it forces 144Hz on my monitor, since it's in list of supported frequencies, but given that the KVM doesn't support it, it blanks out my monitor.  

The fullscreen mode should read the current refresh rate set on the monitor and use that, instead of using the highest refresh rate available.

It actually gets the correct settings, just doesn't use them:
![image](https://github.com/axmolengine/axmol/assets/8603230/04c3417f-6ac4-4d57-b347-a6dfae357a2b)

As you can see, it has "refreshRate = (int) 120" in there.

The monitor blanks out on the call to "createWindow`:
![image](https://github.com/axmolengine/axmol/assets/8603230/6ce97a48-92d5-42bc-afba-1853aecf7682)

In `glfwCreateWindow`:

`window->videoMode.refreshRate = _glfw.hints.refreshRate;`

where `_glfw.hints.refreshRate = -1`

So I'm assuming a call to `glfwWindowHint(GLFW_REFRESH_RATE, videoMode.refreshRate);` is missing in `GLViewImpl::initWithtFullScreen()`.

**Reply by rh101** [reply]
Better fix seems to be just calling the 3 parameter version of `initWithFullScreen` from the other overload, since it's practically the same code.  It compiles fine on Windows, but just doesn't want to compile on MacOS:

![image](https://github.com/axmolengine/axmol/assets/8603230/6ba6c384-34c3-4320-b71e-dc0552c03b1e)

I could only get it to compile by explicitly implementing it all:
```
bool GLViewImpl::initWithFullScreen(std::string_view viewName)
{
    // Create fullscreen window on primary monitor at its current video mode.
    _monitor = glfwGetPrimaryMonitor();
    if (nullptr == _monitor)
        return false;

    const GLFWvidmode* videoMode = glfwGetVideoMode(_monitor);

    // These are soft constraints. If the video mode is retrieved at runtime, the resulting window and context should
    // match these exactly. If invalid attribs are passed (eg. from an outdated cache), window creation will NOT fail
    // but the actual window/context may differ.
    glfwWindowHint(GLFW_REFRESH_RATE, videoMode->refreshRate);
    glfwWindowHint(GLFW_RED_BITS, videoMode->redBits);
    glfwWindowHint(GLFW_BLUE_BITS, videoMode->blueBits);
    glfwWindowHint(GLFW_GREEN_BITS, videoMode->greenBits);

    return initWithRect(viewName, ax::Rect(0, 0, (float)videoMode->width, (float)videoMode->height), 1.0f, false);
}
```

That fixes the issue, so it sets the correct monitor frequency.

**Reply by rh101** [reply]
@halx99 Do you want me to submit a PR for the refresh rate issue?

**Reply by halx99** [reply]
sure, why not

**Reply by IcemarkUK** [reply]
I don't think so no. You need sandbox for uploading to the AppStore and Hardened Runtime for Notorization.

**Reply by IcemarkUK** [reply]
I only set the 14.2 because  I was testing on my laptop and I was getting bugged that it kept not selecting the MacOS SDK which on that machine was 14.2 - setting the option there seemed to work fine.

**Comment by halx99** [comment]
> build success, but when I run it:
> 
> ![image](https://private-user-images.githubusercontent.com/6977319/304283034-fc6bca96-9bcc-4e22-a3b0-7d6a8004af3a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3OTM4NzMsIm5iZiI6MTcwNzc5MzU3MywicGF0aCI6Ii82OTc3MzE5LzMwNDI4MzAzNC1mYzZiY2E5Ni05YmNjLTRlMjItYTNiMC03ZDZhODAwNGFmM2EucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTNUMDMwNjEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTExN2NhYzE1Y2YxY2UzOWI4ZTIyZGVmY2RjZDI5ZTdmZjBmOGJkMjIzZmVhYTc1MGZhNTY3YjgyNGQwOTRjNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.LLidLSQHelJWCWMs-qhxWMWw1OKr5G-E4bHB3oMqQ3Y)
> 
> NOTE: I also enable `-DTESTS=YES`

OK, I reproduce it when remove `-DTESTS=YES`

**Reply by halx99** [reply]
track issue: https://github.com/yasio/yasio/issues/441

**Reply by halx99** [reply, marked as answer]
Fixed by https://github.com/axmolengine/axmol/commit/c7669aac5a5c7c732f5fc93b267e9a7f04b241d2

**Reply by IcemarkUK** [reply]
The -DTESTS option just builds a version that runs unit tests - which I haven't built during this migration.

**Reply by IcemarkUK** [reply]
Many many thanks!!
